store/engine/embedded/raft.go

```diff
@@ -0,0 +1,550 @@
+		 }
+ 
+		 w, err := wal.Create(zap.NewExample(), rc.walDir, nil)
+		 if err != nil {
+			 logger.Get().With(zap.Error(err)).Fatal("Create WAL error")
+		 }
+		 _ = w.Close()
+	 }
+ 
+	 walSnap := walpb.Snapshot{}
+	 if snapshot != nil {
+		 walSnap.Index, walSnap.Term = snapshot.Metadata.Index, snapshot.Metadata.Term
+	 }
+	 logger.Get().Sugar().Infof("Loading WAL at term %d and index %d", walSnap.Term, walSnap.Index)
+	 w, err := wal.Open(logger.Get(), rc.walDir, walSnap)
+	 if err != nil {
+		 logger.Get().With(zap.Error(err)).Fatal("Error loading WAL")
+	 }
+ 
+	 return w
+ }
+ 
+ // replayWAL replays WAL entries into the raft instance.
+ func (rc *raftNode) replayWAL() *wal.WAL {
+	 logger.Get().Sugar().Infof("replaying WAL of member %d", rc.id)
+	 snapshot := rc.loadSnapshot()
+	 w := rc.openWAL(snapshot)
+	 _, st, ents, err := w.ReadAll()
+	 if err != nil {
+		 logger.Get().With(zap.Error(err)).Fatal("Failed to read WAL")
+	 }
+	 rc.raftStorage = raft.NewMemoryStorage()
+	 if snapshot != nil {
+		 _ = rc.raftStorage.ApplySnapshot(*snapshot)
+	 }
+	 _ = rc.raftStorage.SetHardState(st)
+ 
+	 // append to storage so raft starts at the right place in log
+	 _ = rc.raftStorage.Append(ents)
+ 
+	 return w
+ }
+ 
+ func (rc *raftNode) writeError(err error) {
+	 rc.stopHTTP()
+	 close(rc.commitC)
+	 rc.errorC <- err
+	 close(rc.errorC)
+	 rc.node.Stop()
+ }
+ 
+ func (rc *raftNode) startRaft() {
+	 if !fileutil.Exist(rc.snapDir) {
+		 if err := os.Mkdir(rc.snapDir, 0750); err != nil {
+			 logger.Get().With(zap.Error(err)).Fatal("Cannot create dir for snapshot")
+		 }
+	 }
+	 rc.snapshotter = snap.New(logger.Get(), rc.snapDir)
+ 
+	 oldwal := wal.Exist(rc.walDir)
+	 rc.wal = rc.replayWAL()
+ 
+	 // signal replay has finished
+	 rc.snapshotterReady <- rc.snapshotter
+ 
+	 rpeers := make([]raft.Peer, len(rc.peers))
+	 for i := range rpeers {
+		 rpeers[i] = raft.Peer{ID: uint64(i + 1)}
+	 }
+	 c := &raft.Config{
+		 ID:                        uint64(rc.id),
+		 ElectionTick:              10,
+		 HeartbeatTick:             1,
+		 Storage:                   rc.raftStorage,
+		 MaxSizePerMsg:             1024 * 1024,
+		 MaxInflightMsgs:           256,
+		 MaxUncommittedEntriesSize: 1 << 30,
+	 }
+ 
+	 if oldwal || rc.join {
+		 rc.node = raft.RestartNode(c)
+	 } else {
+		 rc.node = raft.StartNode(c, rpeers)
+	 }
+ 
+	 rc.transport = &rafthttp.Transport{
+		 Logger:      logger.Get(),
+		 ID:          types.ID(rc.id),
+		 ClusterID:   0x1000,
+		 Raft:        rc,
+		 ServerStats: stats.NewServerStats("", ""),
+		 LeaderStats: stats.NewLeaderStats(zap.NewExample(), strconv.Itoa(rc.id)),
+		 ErrorC:      make(chan error),
+ 
+		 DialRetryFrequency: 1,
+	 }
+ 
+	 if err := rc.transport.Start(); err != nil {
+		 logger.Get().With(zap.Error(err)).Panic("Failed to start raft http server")
+	 }
+	 for i := range rc.peers {
+		 if i+1 != rc.id {
+			 rc.transport.AddPeer(types.ID(i+1), []string{rc.peers[i]})
+		 }
+	 }
+ 
+	 go rc.serveRaft()
```

Comment:
Should wait for those two routines while stopping

Reply:
`stopCh` would **asychrnously** stop `serveRaft` and `serveChannels`.